base64ct/src/lib.rs

```diff
@@ -202,12 +178,22 @@ pub fn decode<'a>(src: &str, dst: &'a mut [u8], padded: bool) -> Result<&'a [u8]
 }
 
 /// Decode unpadded Base64-encoded string in-place.
-// TODO(tarcieri): support for padded Base64
-pub fn decode_in_place_unpadded(buf: &mut [u8]) -> Result<&[u8], InvalidEncodingError> {
+pub fn decode_in_place(buf: &mut [u8], padded: bool) -> Result<&[u8], InvalidEncodingError> {
     // TODO: eliminate unsafe code when compiler will be smart enough to
     // eliminate bound checks, see: https://github.com/rust-lang/rust/issues/80963
     let mut err: isize = 0;
-    let full_chunks = buf.len() / 4;
+    let dlen = decoded_len(&buf, padded);
+    let full_chunks = if padded {
+        let q = (buf.len() - (buf.len() % 4)) / 4;
+
+        if q > 0 {
+            q - 1
+        } else {
+            q
+        }
+    } else {
+        buf.len() / 4
+    };
```

Comment:
This seems like a pretty gross way to implement this calculation and I'm not fully confident it's correct. Would love suggestions for an alternative.

Reply:
But why do you need such complicated computation instead of simply using `(buf.len() / 4).checked_sub(1).unwrap_or(0)`?

Reply:
Nice! Thank you, that's much more elegant.

I think that will work but (either way) needs a check that `buf.len()` is a multiple of 4.

Reply:
Updated with your suggestion

Reply:
Ooh: clippy recommends `saturating_sub(1)`. Will keep that in mind for the future.

Reply:
Heh, I think it's the first time when I encountered a use for a saturating method. :)

BTW I wonder if adding the following code to the begging of decoding functions can be considered as a violation of const-time properties:
```
if padded && buf.len() % 4 != 0 {
   return Err(..);
}
```
Otherwise the length check has to be done as part of the `decode_3bytes_padded` function.

Reply:
As you mentioned, I think that's fine because it's only variable time in respect to length.

However, there is one part of the code which is data-dependent, which is this loop:

https://github.com/RustCrypto/utils/blob/941e379/base64ct/src/lib.rs#L285-L288

I should add a TODO to fix that (and possibly a GitHub issue)

Reply:
Added a TODO as part of this PR: https://github.com/RustCrypto/utils/pull/227/files#diff-42d97c4412df722211f6a53e6d276f3d9e7f1f6d33a9e13e9ce94ea6003efd52R278-R279